### PR TITLE
[data] Misc: add_column() and allow specifying decoding error handling in from_text()

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This adds some utility functions to make it easier to manipulate structured data in Datasets. While in principle you can already do this with map_batches, this makes it a little easier to test things out for development.

TODO:
- [ ] docs
- [ ] tests

### CODE REVIEW

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Review by clarkzinzow — COMMENTED** [review]
Thanks for opening this RFC!

There are a few high-level things that I'm concerned about:
1. The `with_column()` column adding/mutating API isn't vectorized and is a pretty thin layer over `map_batches()`, so it's sure to be much slower than the `map_batches()` equivalent without a large improvement in UX.
2. The fact that each of these add/mutate/select/drop operations results in a separate `map_batches()` execution stage will make it much less efficient than the single `map_batches()` alternative.

For a concrete before/after comparison, here's the preprocessing stages from the big data training CUJ, along with an added bit to drop columns.

## Before RFC

```python
        def batch_transformer(df: pd.DataFrame):
            # Drop columns.
            df = df.drop(columns["bad_col_1", "bad_col_2"])

            # Drop nulls.
            df = df.dropna(subset=["nullable_feature"])

            # Add new column.
            df["new_col"] = (
                df["feature_1"] - 2 * df["feature_2"] + df["feature_3"]
            ) / 3.0

            # Transform column.
            df["feature_1"] = 2.0 * df["feature_1"] + 0.1

            return df

        ds = ds.map_batches(batch_transformer, batch_format="pandas")
```

## After RFC

```python
        ds = ds.drop_columns("bad_col_1", "bad_col_2") \
            .with_column(
                "new_col", lambda r: (r["feature_1"] - 2 * r["feature_2"] + r["feature_3"]) / 3.0) \
            .with_column("feature_1", lambda r: 2.0 * r["feature_1"] + 1) \
            .map_batches(lambda df: df.dropna(subset=["nullable_feature"]), batch_format="pandas")
```

I don't think that the latter is a big improvement in UX, and it will probably end up being a good bit slower. 🤔 

What do you think?

**Comment by ericl** [comment]
Hmm, what if we made it vectorized by default (so you would have to return a column array)?

File: python/ray/data/dataset.py [hunk 1]
```diff
@@ -180,6 +180,100 @@ def transform(block: Block) -> Iterable[Block]:
...
+        def process_batch(batch):
+            block = BlockAccessor.for_block(batch)
+            df = block.to_pandas()
+            for col in cols:
+                del df[col]
+            return df
+
+        return self.map_batches(process_batch)
+
+    def keep_columns(self, *cols: str) -> "Dataset[T]":
```
[the remark this answers is not in the document]
Comment (clarkzinzow) [reply python/ray/data/dataset.py:252]: We should also encourage column selection at read-time instead of via a mapper task!

**Comment by clarkzinzow** [comment]
>Hmm, what if we made it vectorized by default (so you would have to return a column array)?

That's more reasonable to me. Then there's just the multiple stages of batch mapping that's unfortunate. Once we have a lazy compute model, we could automatically fuse these together.

**Comment by ericl** [comment]
@clarkzinzow ptal

File: python/ray/data/dataset.py [hunk 2]
```diff
@@ -287,6 +287,55 @@ def transform(block: Block) -> Iterable[Block]:
...
+        This is only supported for datasets convertible to pandas format.
+        A function generating the new column values given the batch in pandas
+        format must be specified.
+
+        This is a convenience wrapper over ``.map_batches()``.
+
+        Examples:
+            >>> ds = ray.data.range_arrow(100)
+            >>> # Add a new column equal to value * 2.
+            >>> ds = ds.add_column("new_col", lambda r: r["value"] * 2)
```
Comment (jjyao) [inline python/ray/data/dataset.py:309]: nit: `r` -> `df` to make it clearer that the input is a data frame?

File: python/ray/data/dataset.py [hunk 3]
```diff
@@ -287,6 +287,55 @@ def transform(block: Block) -> Iterable[Block]:
...
+                records in pandas format.
+            compute: The compute strategy, either "tasks" (default) to use Ray
+                tasks, or "actors" to use an autoscaling Ray actor pool.
+            ray_remote_args: Additional resource requirements to request from
+                ray (e.g., num_gpus=1 to request GPUs for the map tasks).
+        """
+
+        def process_batch(batch):
+            block = BlockAccessor.for_block(batch)
+            df = block.to_pandas()
```
Comment (jjyao) [inline python/ray/data/dataset.py:328]: why do we need to convert to pandas again if  batch_format is pandas?

File: python/ray/data/dataset.py [hunk 4]
```diff
@@ -287,6 +287,55 @@ def transform(block: Block) -> Iterable[Block]:
         blocks = compute.apply(transform, ray_remote_args, self._blocks)
         return Dataset(blocks, self._epoch, stats_builder.build(blocks))
 
+    def add_column(
```
Comment (jjyao) [inline python/ray/data/dataset.py:290]: Hmm I'm not sure about the benefit-cost ratio: given we support pandas batch format, it should be trivial for users to just use map_batch?

Comment (ericl) [reply python/ray/data/dataset.py:290]: It's much more verbose when using map_batch (1 loc vs 5-6). I think this is quite confusing for users, who typically expect more streamlined utility functions.

File: python/ray/data/dataset.py (hunk 2 above)
Comment (ericl) [reply python/ray/data/dataset.py]: done

File: python/ray/data/dataset.py (hunk 3 above)
Comment (ericl) [reply python/ray/data/dataset.py]: done

**Review by jjyao — APPROVED** [review]

**Review by scv119 — APPROVED** [review]

**Pull request merged**